src/api/bulk2.ts

```diff
@@ -0,0 +1,1000 @@
+   *
+   * @param interval Polling interval in milliseconds
+   * @param timeout Polling timeout in milliseconds
+   * @returns {Promise<Record[]>} A promise that resolves when the job finished being processed.
+   */
+  async poll(
+    interval: number = this.pollingOptions.pollInterval,
+    timeout: number = this.pollingOptions.pollTimeout,
+  ): Promise<void> {
+    const jobId = getJobId(this.jobInfo);
+    const startTime = Date.now();
+
+    this.logger.debug(`Job ID: ${jobId}`);
+    this.logger.debug(
+      `Polling options: timeout:${timeout}ms | interval: ${interval}ms.`,
+    );
+
+    while (startTime + timeout > Date.now()) {
+      try {
+        const res = await this.check();
+        switch (res.state) {
+          case 'Aborted':
+            throw new Error('Job has been aborted');
+          case 'UploadComplete':
+          case 'InProgress':
+            await delay(interval);
+            break;
+          case 'Failed':
+            // unlike ingest jobs, the API doesn't return an error msg:
+            // https://developer.salesforce.com/docs/atlas.en-us.api_asynch.meta/api_asynch/query_get_one_job.htm
+            this.logger.debug(res);
+            throw new Error('Query job failed to complete');
+          case 'JobComplete':
+            this.emit('jobcomplete');
+            return;
+        }
+      } catch (err) {
+        this.emit('error', err);
+        throw err;
+      }
+    }
+
+    const timeoutError = new JobPollingTimeoutError(
+      `Polling timed out after ${timeout}ms. Job Id = ${jobId}`,
+      jobId,
+    );
+    this.emit('error', timeoutError);
+    throw timeoutError;
+  }
+
+  /**
+   * Check the latest job status
+   *
+   * @returns {Promise<QueryJobInfoV2>} job information.
+   */
+  async check(): Promise<QueryJobInfoV2> {
+    try {
+      const jobInfo = await this.createQueryRequest<QueryJobInfoV2>({
+        method: 'GET',
+        path: `/${getJobId(this.jobInfo)}`,
+        responseType: 'application/json',
+      });
+      this.jobInfo = jobInfo;
+      this.logger.debug(this.jobInfo);
+      return jobInfo;
+    } catch (err) {
+      this.emit('error', err);
+      throw err;
+    }
+  }
+
+  /**
+   * Get the results for a query job as a record stream
+   *
+   * This method assumes the job finished being processed
+   * @returns {RecordStream} - Record stream, convertible to a CSV data stream
+   */
+  public async result(): Promise<Parsable<Record>> {
+    const resultStream = new Parsable();
+    const resultDataStream = resultStream.stream('csv');
+
+    const resultsPath = `/${getJobId(this.jobInfo)}/results`;
+
+    while (this.locator !== 'null') {
+      const resPromise = this.createQueryRequest({
+        method: 'GET',
+        path: this.locator
+          ? `${resultsPath}?locator=${this.locator}`
+          : resultsPath,
+        headers: {
+          Accept: 'text/csv',
+        },
+      });
+
+      resPromise.stream().pipe(resultDataStream);
```

Comment:
this will attach the writable data stream to the readable inside the HTTP API module:
https://github.com/jsforce/jsforce/blob/588c6cf986650fb68f9b41314d5c48a9a5c5d478/src/http-api.ts#L63-L66

Reply:
how it works:
https://github.com/jsforce/jsforce/blob/2.0/src/util/promise.ts